foreman/resource_foreman_host.go

```diff
@@ -519,42 +519,46 @@ func resourceForemanHostCreate(d *schema.ResourceData, meta interface{}) error {
 
 	enablebmc := d.Get("enable_bmc").(bool)
 
+	var powerCmds []interface{}
 	// If enable_bmc is true, perform required power off, pxe boot and power on BMC functions
 	if enablebmc {
 		log.Debugf("Calling BMC Reboot/PXE Functions")
 		// List of BMC Actions to perform
-		bmcCmds := []interface{}{
-			api.BMCPower{
-				PowerAction: api.BmcPowerOff,
+		powerCmds = []interface{}{
+			api.Power{
+				PowerAction: api.PowerOff,
 			},
 			api.BMCBoot{
-				Device: api.BmcBootPxe,
+				Device: api.BootPxe,
 			},
-			api.BMCPower{
-				PowerAction: api.BmcPowerOn,
+			api.Power{
+				PowerAction: api.PowerOn,
 			},
 		}
-
-		// Loop through each of the above BMC Operations and execute.
-		// In the event fo any failure, exit with error
-		for _, cmd := range bmcCmds {
-			sendBMCErr := client.SendBMCCommand(createdHost, cmd, hostRetryCount)
-			if sendBMCErr != nil {
-				return sendBMCErr
-			}
-			// Sleep for 3 seconds between chained BMC calls
-			duration := time.Duration(3) * time.Second
-			time.Sleep(duration)
-		}
-		// When the BMC Operations succeed, set the `bmc_success` key to true.
-		d.Set("bmc_success", true)
-		// Set the `bmc_success` key as successful in partial mode
-		d.SetPartial("bmc_success")
-
 	} else {
```

Comment:
I agree with the generalization of power operations above to support other providers but our intended usage of this particular Boolean is to provide a toggle for all BMC operations for the defined host.  With this change, A `false` state for this property would still trigger a PowerOn BMC operation which may be undesired.  For baremetal hosts, the chain of `Power off/Set to PXE boot/Power on` with a brief delay was the most reliable approach we found for both net new hosts and rebuilds of existing hosts. We do not use foreman to provision VMs in our environment but I'm guessing this may not be the case when provisioning them. 

One arguably cleaner way I can think to accomplish this might be to expose the `powerCmds` power/boot operation chain as an optional, defaulted ordered list of strings to the host resource so we can override the operations to our hearts content in-module.  We would probably want this attribute to NOT trigger a host rebuild.